### DISCUSSION TITLE
Failing `clang-tidy` checks

### DISCUSSION THREAD

**Discussion by bcaddy** [cholla-hydro/cholla#217] [Ideas]
First column is total warnings, second column is percent of total failures, third column is the check

[Glossary of checks in clang-tidy 15](https://releases.llvm.org/15.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/list.html)

```
3623  16.667 readability-identifier-naming
3048  14.022 readability-identifier-length
1482   6.818 cppcoreguidelines-avoid-non-const-global-variables
1317   6.059 cppcoreguidelines-init-variables
1132   5.208 readability-magic-numbers
1132   5.208 cppcoreguidelines-avoid-magic-numbers
1122   5.162 cppcoreguidelines-pro-bounds-pointer-arithmetic
 742   3.414 cppcoreguidelines-macro-usage
 487    2.24 readability-braces-around-statements
 484   2.227 cppcoreguidelines-narrowing-conversions
 484   2.227 bugprone-narrowing-conversions
 461   2.121 misc-const-correctness
 404   1.859 modernize-macro-to-enum
 335   1.541 hicpp-braces-around-statements
 281   1.293 readability-isolate-declaration
 266   1.224 clang-analyzer-deadcode.DeadStores
 257   1.182 google-readability-casting
 228   1.049 modernize-use-nodiscard
 191   0.879 llvm-header-guard
 191   0.879 bugprone-easily-swappable-parameters
 171   0.787 google-runtime-int
 166   0.764 modernize-use-using
 153   0.704 hicpp-no-array-decay
 153   0.704 cppcoreguidelines-pro-bounds-array-to-pointer-decay
 152   0.699 google-readability-braces-around-statements
 145   0.667 hicpp-vararg
 145   0.667 cppcoreguidelines-pro-type-vararg
 124    0.57 misc-unused-parameters
 117   0.538 modernize-avoid-c-arrays
 117   0.538 hicpp-avoid-c-arrays
 117   0.538 cppcoreguidelines-avoid-c-arrays
 111   0.511 cert-err33-c
 108   0.497 llvm-include-order
 104   0.478 modernize-use-override
 104   0.478 hicpp-use-override
 104   0.478 cppcoreguidelines-explicit-virtual-functions
  98   0.451 cert-dcl51-cpp
  98   0.451 cert-dcl37-c
  98   0.451 bugprone-reserved-identifier
  94   0.432 cppcoreguidelines-owning-memory
  88   0.405 bugprone-implicit-widening-of-multiplication-result
  82   0.377 hicpp-no-malloc
  82   0.377 cppcoreguidelines-no-malloc
  76    0.35 cppcoreguidelines-pro-type-cstyle-cast
  68   0.313 modernize-use-nullptr
  61   0.281 hicpp-explicit-conversions
  61   0.281 google-explicit-constructor
  59   0.271 modernize-deprecated-headers
  59   0.271 hicpp-deprecated-headers
  58   0.267 cert-err34-c
  57   0.262 hicpp-special-member-functions
  57   0.262 cppcoreguidelines-special-member-functions
  53   0.244 modernize-redundant-void-arg
  53   0.244 llvm-namespace-comment
  50    0.23 google-readability-namespace-comments
  43   0.198 misc-non-private-member-variables-in-classes
  42   0.193 readability-non-const-parameter
  41   0.189 readability-make-member-function-const
  41   0.189 concurrency-mt-unsafe
  35   0.161 cppcoreguidelines-prefer-member-initializer
  30   0.138 hicpp-use-nullptr
  25   0.115 readability-implicit-bool-conversion
  25   0.115 readability-else-after-return
  25   0.115 llvm-else-after-return
  20   0.092 readability-redundant-control-flow
  20   0.092 cppcoreguidelines-non-private-member-variables-in-classes
  19   0.087 clang-analyzer-security.insecureAPI.strcpy
  17   0.078 readability-function-cognitive-complexity
  16   0.074 readability-inconsistent-declaration-parameter-name
  15   0.069 hicpp-member-init
  15   0.069 cppcoreguidelines-pro-type-member-init
  14   0.064 bugprone-macro-parentheses
  13    0.06 modernize-use-noexcept
  13    0.06 hicpp-use-noexcept
  12   0.055 google-readability-todo
  11   0.051 readability-convert-member-functions-to-static
  10   0.046 readability-suspicious-call-argument
  10   0.046 misc-confusable-identifiers
  10   0.046 cppcoreguidelines-pro-bounds-constant-array-index
  10   0.046 clang-analyzer-core.CallAndMessage
   9   0.041 hicpp-signed-bitwise
   9   0.041 clang-analyzer-core.UndefinedBinaryOperatorResult
   7   0.032 cert-err58-cpp
   4   0.018 performance-unnecessary-value-param
   4   0.018 google-build-namespaces
   4   0.018 cppcoreguidelines-virtual-class-destructor
   4   0.018 cert-dcl59-cpp
   3   0.014 readability-simplify-boolean-expr
   3   0.014 readability-const-return-type
   3   0.014 performance-inefficient-vector-operation
   3   0.014 modernize-use-auto
   3   0.014 hicpp-use-auto
   3   0.014 cert-msc51-cpp
   3   0.014 cert-msc32-c
   3   0.014 cert-env33-c
   3   0.014 bugprone-string-integer-assignment
   2   0.009 performance-for-range-copy
   2   0.009 performance-faster-string-find
   2   0.009 modernize-use-default-member-init
   2   0.009 modernize-loop-convert
   2   0.009 hicpp-multiway-paths-covered
   2   0.009 google-build-using-namespace
   2   0.009 bugprone-integer-division
   2   0.009 bugprone-branch-clone
   2   0.009 bugprone-assignment-in-if-condition
   1   0.005 modernize-use-emplace
   1   0.005 modernize-use-bool-literals
   1   0.005 hicpp-use-emplace
   1   0.005 clang-analyzer-valist.Uninitialized
   1   0.005 clang-analyzer-optin.performance.Padding
   1   0.005 clang-analyzer-core.uninitialized.Assign
   1   0.005 cert-str34-c
   1   0.005 cert-dcl50-cpp
   1   0.005 bugprone-signed-char-misuse
```

**Comment by alwinm** [comment]
TODO for Bob: short tutorial of how to run clang tidy to extract the actual instances associated with one of the errors

**Reply by bcaddy** [reply]
tldr. The `.clang-tidy` file has a list of all disabled checks. See the header comment in that file for the ordering on enabling/disabling checks then move the appropriate check into the appropriate place.

**Reply by bcaddy** [reply]
If you want to run all the enabled checks then type

`make tidy` 

~If you only want to run a handful of checks (or pass arguments to `clang-tidy`)~

~`make tidy CLANG_TIDY_ARGS="-checks=YOUR-CHECKS-HERE"`~

Edit: See below for correct info

**Reply by bcaddy** [reply]
The above info isn't quite correct. I've added clang-tidy documentation to the [Making a Pull Request](https://github.com/cholla-hydro/cholla/wiki/Making-a-Pull-Request) section of the wiki. It includes how to run specific checks and how to run on specific files

**Comment by alwinm** [comment]
The above list, reformatted as a task list. 

- [ ] 3623  16.667 readability-identifier-naming
  - [ ] variables
  - [ ] functions
  - [x] namespaces 
  - [x] macros
  - [x] typdefs
  - [x] type aliases
  - [x] enums
  - [ ] constants
  - [ ] constant prefix
  - [ ] global prefix
  - [x] class name
  - [ ] private suffix
- [ ] 3048  14.022 readability-identifier-length
- [ ] 1482   6.818 cppcoreguidelines-avoid-non-const-global-variables
- [ ] 1317   6.059 cppcoreguidelines-init-variables
- [ ] 1132   5.208 readability-magic-numbers
- [ ] 1132   5.208 cppcoreguidelines-avoid-magic-numbers
- [ ] 1122   5.162 cppcoreguidelines-pro-bounds-pointer-arithmetic
- [ ]  742   3.414 cppcoreguidelines-macro-usage
- [x]  487    2.24 readability-braces-around-statements
- [ ]  484   2.227 cppcoreguidelines-narrowing-conversions **(Mostly issues with converting to `int` since we use `int` for all our indexing)**
- [ ]  484   2.227 bugprone-narrowing-conversions **(alias to cppcoreguidelines-narrowing-conversions)**
- [ ]  461   2.121 misc-const-correctness
- [ ]  404   1.859 modernize-macro-to-enum
- [x]  335   1.541 hicpp-braces-around-statements
- [ ]  281   1.293 readability-isolate-declaration
- [ ]  266   1.224 clang-analyzer-deadcode.DeadStores
- [ ]  257   1.182 google-readability-casting
- [ ]  228   1.049 modernize-use-nodiscard
- [ ]  191   0.879 llvm-header-guard
- [x]  191   0.879 bugprone-easily-swappable-parameters **(will not enable. Checks for contiguous arguments with the same type which we will always have many of and refactoring around this would be time consuming with no real benefit)**
- [ ]  171   0.787 google-runtime-int
- [ ]  166   0.764 modernize-use-using
- [ ]  153   0.704 hicpp-no-array-decay
- [ ]  153   0.704 cppcoreguidelines-pro-bounds-array-to-pointer-decay
- [x]  152   0.699 google-readability-braces-around-statements (done, see #246)
- [ ]  145   0.667 hicpp-vararg
- [ ]  145   0.667 cppcoreguidelines-pro-type-vararg
- [x]  124    0.57 misc-unused-parameters (**affected too much by ifdefs to be useful**)
- [ ]  117   0.538 modernize-avoid-c-arrays
- [ ]  117   0.538 hicpp-avoid-c-arrays
- [ ]  117   0.538 cppcoreguidelines-avoid-c-arrays
- [ ]  111   0.511 cert-err33-c
- [ ]  108   0.497 llvm-include-order
- [ ]  104   0.478 modernize-use-override
- [ ]  104   0.478 hicpp-use-override
- [ ]  104   0.478 cppcoreguidelines-explicit-virtual-functions
- [x]   98   0.451 cert-dcl51-cpp **(enabled, alias of bugprone-reserved-identifier)**
- [x]   98   0.451 cert-dcl37-c **(enabled, alias of bugprone-reserved-identifier)**
- [x]   98   0.451 bugprone-reserved-identifier (**Enabled. Can use the `AllowedIdentifiers` option to whitelist identifiers if needed**)
- [ ]   94   0.432 cppcoreguidelines-owning-memory
- [ ]   88   0.405 bugprone-implicit-widening-of-multiplication-result **(Mostly issues with converting to `int` since we use `int` for all our indexing)**
- [x]   82   0.377 hicpp-no-malloc **(maybe fix when people have time to do it)**
- [x]   82   0.377 cppcoreguidelines-no-malloc **(maybe fix when people have time to do it, but there are many mallocs)**
- [ ]   76    0.35 cppcoreguidelines-pro-type-cstyle-cast
- [ ]   68   0.313 modernize-use-nullptr
- [ ]   61   0.281 hicpp-explicit-conversions
- [ ]   61   0.281 google-explicit-constructor
- [ ]   59   0.271 modernize-deprecated-headers
- [ ]   59   0.271 hicpp-deprecated-headers
- [ ]   58   0.267 cert-err34-c
- [ ]   57   0.262 hicpp-special-member-functions
- [ ]   57   0.262 cppcoreguidelines-special-member-functions
- [ ]   53   0.244 modernize-redundant-void-arg
- [ ]   53   0.244 llvm-namespace-comment
- [ ]   50    0.23 google-readability-namespace-comments
- [ ]   43   0.198 misc-non-private-member-variables-in-classes
- [ ]   42   0.193 readability-non-const-parameter
- [ ]   41   0.189 readability-make-member-function-const
- [ ]   41   0.189 concurrency-mt-unsafe
- [ ]   35   0.161 cppcoreguidelines-prefer-member-initializer
- [ ]   30   0.138 hicpp-use-nullptr
- [ ]   25   0.115 readability-implicit-bool-conversion
- [ ]   25   0.115 readability-else-after-return
- [ ]   25   0.115 llvm-else-after-return
- [ ]   20   0.092 readability-redundant-control-flow
- [ ]   20   0.092 cppcoreguidelines-non-private-member-variables-in-classes
- [ ]   19   0.087 clang-analyzer-security.insecureAPI.strcpy
- [ ]   17   0.078 readability-function-cognitive-complexity
- [ ]   16   0.074 readability-inconsistent-declaration-parameter-name
- [ ]   15   0.069 hicpp-member-init
- [ ]   15   0.069 cppcoreguidelines-pro-type-member-init
- [x]   14   0.064 bugprone-macro-parentheses **(fixed and enabled)**
- [x]   13    0.06 modernize-use-noexcept (**all errors are in HDF5**)
- [x]   13    0.06 hicpp-use-noexcept (**all errors are in HDF5**)
- [ ]   12   0.055 google-readability-todo
- [ ]   11   0.051 readability-convert-member-functions-to-static
- [ ]   10   0.046 readability-suspicious-call-argument
- [x]   10   0.046 misc-confusable-identifiers **(fixed and enabled)**
- [x]   10   0.046 cppcoreguidelines-pro-bounds-constant-array-index **(disabled. It's not feasible in GPU code)**
- [ ]   10   0.046 clang-analyzer-core.CallAndMessage **(these are very valid code smells but will take work to refactor.)**
- [x]    9   0.041 hicpp-signed-bitwise **(fixed and enabled)**
- [ ]    9   0.041 clang-analyzer-core.UndefinedBinaryOperatorResult  (**these are very valid code smells but will take work to refactor. See #308**)
- [x]    7   0.032 cert-err58-cpp **(disabled. Minimal gains and it flags every gtest `TEST` macro as failing)**
- [x]    4   0.018 performance-unnecessary-value-param  **(fixed and enabled)**
- [x]    4   0.018 google-build-namespaces **(fixed and enabled)**
- [x]    4   0.018 cppcoreguidelines-virtual-class-destructor (**all errors are in HDF5**)
- [x]    4   0.018 cert-dcl59-cpp **(alias togoogle-build-namespaces )**
- [x]    3   0.014 readability-simplify-boolean-expr **(fixed and enabled)**
- [x]    3   0.014 readability-const-return-type **(enabled, issue no longer present at time of checking)**
- [x]    3   0.014 performance-inefficient-vector-operation **(enabled)**
- [x]    3   0.014 modernize-use-auto  **(fixed and enabled)**
- [x]    3   0.014 hicpp-use-auto **(enabled, alias of modernize-use-auto )**
- [x]    3   0.014 cert-msc51-cpp (**This check wants to remove RNG seeds but we want seeds**)
- [x]    3   0.014 cert-msc32-c (**This check wants to remove RNG seeds but we want seeds**)
- [x]    3   0.014 cert-env33-c (**This check wants to remove RNG seeds but we want seeds**)
- [x]    3   0.014 bugprone-string-integer-assignment **(fixed and enabled)**
- [x]    2   0.009 performance-for-range-copy **(fixed and enabled)**
- [x]    2   0.009 performance-faster-string-find  **(fixed and enabled)**
- [x]    2   0.009 modernize-use-default-member-init  **(fixed and enabled)**
- [x]    2   0.009 modernize-loop-convert  **(fixed and enabled)**
- [x]    2   0.009 hicpp-multiway-paths-covered **(will not enable. Too many use cases where it would flag good code for not much benefit)**
- [x]    2   0.009 google-build-using-namespace **(fixed and enabled)**
- [x]    2   0.009 bugprone-integer-division **(fixed and enabled)**
- [x]    2   0.009 bugprone-branch-clone **(fixed and enabled)**
- [x]    2   0.009 bugprone-assignment-in-if-condition **(fixed and enabled)**
- [x]    1   0.005 **modernize-use-emplace **(fixed and enabled)**
- [x]    1   0.005 modernize-use-bool-literals **(fixed and enabled)**
- [x]    1   0.005 hicpp-use-emplace **(alias to modernize-use-emplace)**
- [x]    1   0.005 clang-analyzer-valist.Uninitialized **(fixed and enabled)**
- [x]    1   0.005 clang-analyzer-optin.performance.Padding (**all errors are in HDF5**)
- [x]    1   0.005 clang-analyzer-core.uninitialized.Assign **(fixed and enabled)**
- [x]    1   0.005 cert-str34-c **(alias to bugprone-signed-char-misuse)**
- [x]    1   0.005 cert-dcl50-cpp **(fixed and enabled)**
- [x]    1   0.005 bugprone-signed-char-misuse (would need to replace or have clang ignore djb2_hash in MPI_Comm_Node.c)
- [x]   0  Alwin Dummy Example (**edited with comment as to why we aren't looking at this. It's a dummy example not a real clang error**)

**Reply by alwinm** [reply]
What I envision is that we can use this list to check off things that have been looked at, though ideally it would let someone else add comments since letting only 1 person comment is not great. I do think a linear format for this makes sense since it is probably easier to look at the clang-tidy issues at the bottom of the list first with fewer violations.

**Reply by alwinm** [reply]
I just checked and it looks like Maintainers can edit comments so you can edit the above list, I believe.

**Reply by bcaddy** [reply]
This looks great. Then people can grab what they want, create a card, and work on it.

If a particular check needs more discussion it can can be added as it's own thread in this Discussion and we can discuss it.

**Comment by alwinm** [comment]
Output from the python script I made to parse clang-tidy output 

[cert-dcl50-cpp] 1
io.cpp:3567:5: 1

[clang-analyzer-valist.Uninitialized] 1
io.cpp:3577:12: 1

[hicpp-use-emplace 1
main_tests.cpp:90:21: 1

modernize-use-emplace] 1
main_tests.cpp:90:21: 1

[clang-analyzer-core.uninitialized.Assign] 1
mpi_routines.cpp:538:19: 1

[modernize-use-bool-literals] 1
mpi_routines.cpp:824:10: 1

[bugprone-signed-char-misuse 1
MPI_Comm_node.c:61:15: 1

cert-str34-c] 1
MPI_Comm_node.c:61:15: 1

[clang-analyzer-optin.performance.Padding] 1
H5ACpublic.h:447:16: 1

[google-readability-namespace-comments 1
H5Classes.h:45:1: 1

llvm-namespace-comment] 1
H5Classes.h:45:1: 1

[modernize-loop-convert] 2
global.cpp:109:3: 1
reduction_utilities_tests.cu:48:3: 1

[google-build-using-namespace] 2
initial_conditions.cpp:25:1: 1
io.cpp:26:1: 1

[bugprone-branch-clone] 2
initial_conditions.cpp:644:46: 1
testing_utilities.cpp:64:32: 1

[hicpp-multiway-paths-covered] 2
mpi_boundaries.cpp:944:5: 1
mpi_routines.cpp:967:3: 1

[bugprone-integer-division] 2
disk_ICs.cpp:171:40: 1
disk_ICs.cpp:174:29: 1

[bugprone-assignment-in-if-condition] 2
mpi_routines.cpp:186:9: 1
mpi_routines.cpp:190:9: 1

[performance-faster-string-find] 2
system_tester.cpp:419:47: 1
system_tester.cpp:422:71: 1

[modernize-use-default-member-init] 2
ran.h:9:13: 1
ran.h:9:16: 1

[performance-inefficient-vector-operation] 3
disk_galaxy.h:190:7: 3

[readability-simplify-boolean-expr] 3
boundary_conditions.cpp:158:9: 1
main_tests.cpp:132:23: 1
main_tests.cpp:138:38: 1

[cert-msc32-c 3
initial_conditions.cpp:959:3: 1
io.cpp:2006:3: 1
reduction_utilities_tests.cu:44:16: 1

cert-msc51-cpp] 3
initial_conditions.cpp:959:3: 1
io.cpp:2006:3: 1
reduction_utilities_tests.cu:44:16: 1

[performance-for-range-copy] 3
system_tester.cpp:110:13: 1
system_tester.cpp:253:13: 1
system_tester.cpp:633:15: 1

[cert-env33-c] 3
system_tester.cpp:322:21: 1
system_tester.cpp:325:23: 1
system_tester.cpp:463:7: 1

[bugprone-string-integer-assignment] 3
testing_utilities.cpp:85:16: 1
testing_utilities.cpp:87:16: 1
testing_utilities.cpp:89:16: 1

[hicpp-use-auto 4
io.cpp:318:5: 1
reduction_utilities.h:126:3: 3

modernize-use-auto] 4
io.cpp:318:5: 1
reduction_utilities.h:126:3: 3

[cppcoreguidelines-virtual-class-destructor] 4
H5Classes.h:21:15: 1
H5Classes.h:22:15: 1
H5Location.h:31:17: 1
H5Object.h:55:17: 1

[readability-const-return-type] 6
DeviceVector.h:265:1: 6

[cert-err58-cpp] 7
disk_galaxy.h:206:28: 3
disk_galaxy.h:208:19: 3
global.cpp:97:30: 1

[clang-diagnostic-unused-command-line-argument] 7
warning: 7

[clang-analyzer-core.UndefinedBinaryOperatorResult] 8
io.cpp:363:16: 1
io.cpp:514:14: 1
plmp_cuda.cu:112:11: 1
ppmc_cuda.cu:143:11: 1
exact_cuda.cu:73:27: 1
hll_cuda.cu:80:27: 1
hllc_cuda.cu:81:27: 1
roe_cuda.cu:82:27: 1

[clang-analyzer-core.CallAndMessage] 10
mpi_boundaries.cpp:408:9: 1
mpi_boundaries.cpp:492:9: 1
mpi_boundaries.cpp:585:9: 1
mpi_boundaries.cpp:670:9: 1
mpi_boundaries.cpp:761:9: 1
mpi_boundaries.cpp:846:9: 1
io.cpp:1853:3: 1
io.cpp:1973:3: 1
io.cpp:3562:3: 1
main.cpp:75:3: 1

[readability-convert-member-functions-to-static] 11
disk_galaxy.h:16:8: 3
initial_conditions.cpp:1492:14: 1
initial_conditions.cpp:1597:14: 1
system_tester.cpp:370:51: 1
system_tester.cpp:383:51: 1
system_tester.cpp:488:36: 1
system_tester.cpp:780:56: 1
hllc_cuda_tests.cu:47:21: 1
hllc_cuda_tests.cu:113:8: 1

[performance-unnecessary-value-param] 11
testing_utilities.h:128:31: 8
testing_utilities.cpp:78:52: 1
testing_utilities.cpp:97:35: 1
testing_utilities.cpp:115:31: 1

[misc-confusable-identifiers] 11
chemistry_gpu.h:53:9: 2
gpu.hpp:342:13: 8
roe_cuda.cu:327:24: 1

[cppcoreguidelines-pro-bounds-constant-array-index] 12
mpi_routines.cpp:409:27: 1
mpi_routines.cpp:409:39: 1
reduction_utilities.h:67:18: 3
reduction_utilities.h:72:49: 3
cuda_boundaries.cu:132:9: 1
error_check_cuda.cu:44:5: 1
error_check_cuda.cu:50:13: 1
error_check_cuda.cu:50:27: 1

[hicpp-use-noexcept 13
H5Exception.h:84:30: 1
H5Exception.h:99:35: 1
H5Exception.h:106:36: 1
H5Exception.h:113:40: 1
H5Exception.h:120:39: 1
H5Exception.h:127:40: 1
H5Exception.h:134:39: 1
H5Exception.h:141:38: 1
H5Exception.h:148:40: 1
H5Exception.h:155:39: 1
H5Exception.h:162:38: 1
H5Exception.h:169:38: 1
H5Exception.h:176:41: 1

modernize-use-noexcept] 13
H5Exception.h:84:30: 1
H5Exception.h:99:35: 1
H5Exception.h:106:36: 1
H5Exception.h:113:40: 1
H5Exception.h:120:39: 1
H5Exception.h:127:40: 1
H5Exception.h:134:39: 1
H5Exception.h:141:38: 1
H5Exception.h:148:40: 1
H5Exception.h:155:39: 1
H5Exception.h:162:38: 1
H5Exception.h:169:38: 1
H5Exception.h:176:41: 1

**Comment by alwinm** [comment]
```py

import os                                                                                                                                                                                                                   
import sys                                                                                                                                                                                                                  
                                                                                                                                                                                                                            
number_of_violations_to_print = 40                                                                                                                                                                                          
                                                                                                                                                                                                                            
counter = {}                                                                                                                                                                                                                
                                                                                                                                                                                                                            
                                                                                                                                                                                                                            
def process(filename):                                                                                                                                                                                                      
    with open(filename,'r') as ofile:                                                                                                                                                                                       
        data = ofile.readlines()                                                                                                                                                                                            
                                                                                                                                                                                                                            
    for line in data:                                                                                                                                                                                                       
        if 'warning:' in line:                                                                                                                                                                                              
            ls = line.split()                                                                                                                                                                                               
            code_file_path = ls[0]                                                                                                                                                                                          
            code_file = code_file_path.split('/')[-1]                                                                                                                                                                       
                                                                                                                                                                                                                            
            violation_string = ls[-1]                                                                                                                                                                                       
            violations = violation_string.split(',')                                                                                                                                                                        
                                                                                                                                                                                                                            
            for violation in violations:                                                                                                                                                                                    
                if violation not in counter:                                                                                                                                                                                
                    counter[violation] = {}                                                                                                                                                                                 
                                                                                                                                                                                                                            
                if code_file not in counter[violation]:                                                                                                                                                                     
                    counter[violation][code_file] = 0                                                                                                                                                                       
                                                                                                                                                                                                                            
                counter[violation][code_file] += 1                                                                                                                                                                          
                                                                                                                                                                                                                            
                                                                                                                                                                                                                            
process('tidy_results_cpp.log')                                                                                                                                                                                             
process('tidy_results_c.log')                                                                                                                                                                                               
process('tidy_results_gpu.log')                                                                                                                                                                                             
                                                                                                                                                                                                                            
def get_total_count(key):                                                                                                                                                                                                   
    return sum(counter[key].values())                                                                                                                                                                                       
                                                                                                                                                                                                                            
vios = sorted(list(counter.keys()),key=get_total_count)                                                                                                                                                                     
                                                                                                                                                                                                                            
for i in range(number_of_violations_to_print):                                                                                                                                                                              
    print()                                                                                                                                                                                                                 
    print(vios[i],get_total_count(vios[i]))                                                                                                                                                                                 
    for key in counter[vios[i]]:                                                                                                                                                                                            
        print(key,counter[vios[i]][key])    
```

**Comment by bcaddy** [comment]
About `bugprone-signed-char-misuse`. This shows up in our singular `.c` file which (as far as I can tell) exists solely to implement the  `MPI_Comm_node` function. It looks like this function is used to set the global variable `node` which (again, as far as I can tell) is never used. Is there a problem with just cutting this out entirely? It also means we can get rid of the singular C file and all the associated build complexity that comes with generating a multi-language binary.

@evaneschneider and @alwinm, I'd like your input on this.

**Reply by evaneschneider** [reply]
I agree. I don't think it's used anywhere. I'm fine with deleting that source file and the associated calls in mpi_routines.cpp and simplifying the makefile.

**Reply by bcaddy** [reply]
This is now documented in Issue #253

**Comment by bcaddy** [comment]
The last two `bugprone` checks that we might want to enable are:

- `bugprone-implicit-widening-of-multiplication-result`
- `bugprone-narrowing-conversions` (alias of `cppcoreguidelines-narrowing-conversions`)

For both of these the bulk of the errors are because we use `int` and not `size_t` for indexing arrays/sizes of things. Fixing this is not that hard but would require a lot of small changes. Are there any thoughts on this or is it just something we could work on during hack sessions?

**Reply by alwinm** [reply]
After dealing with i32 vs usize on Rust (where Rust forces that you do make these conversions explicit or it does not compile), my opinion is that turning some int into size_t will result in more things being needed to turned from int to size_t and tbh it is going to be more effort than it's worth. 

Basically at some point you would want to identify the narrowest place where you could make the conversions explicit with (int) or (size_t) but that's more work than it's worth. 

I recommend we do not fix until it actually can or does cause a bug.

**Reply by alwinm** [reply]
In the case that this fix goes ahead, the major trap I've run into and we should try to avoid is that size_t is unsigned so doesn't conveniently store negative values. 

Here's a simple example
```
#include <cstddef>
#include <cassert>
int main() {
 size_t a = 0 - 10;
 assert(a > 10);
 assert(a > 0);

}
```

This means that fixes will need to be done thoughtfully. One approach is to use ptrdiff_t instead of size_t everywhere.

**Reply by bcaddy** [reply]
I should clarify. My question was less "should we invest the dev time into switching to `size_t`?" and more "Is it worth investing the time to enable the narrowing/implicit conversion checks?" since improper conversions can lead to subtle and hard to find bugs and if we can automate checking for them that would be great. There was also quite a few `double` to `float` conversions floating around so this is possibly a current issue.

In terms of capacity of `int`. With 11 fields (hydro, mhd, dust, DE, and one scalar) that limits us to 580^3 cells per gpu. While this isn't an issue right now it very well could be on future hardware if VRAM continues growing at the current rate. 

IMO being forced to change all of our length variables to `size_t` isn't a bad thing since it will make the code clearer about what is what. We'll likely have to do some explicit conversions at some point but hopefully they will be minimal.

`ptrdiff_t` and `size_t` are not interchangeable. On 64 bit systems they're 64 bit signed and unsigned integers respectively which means that `ptrdiff_t` only has half the positive range of `size_t` and therefore doesn't have the capacity to index any allocatable array.

**Reply by alwinm** [reply]
> ptrdiff_t and size_t are not interchangeable. On 64 bit systems they're 64 bit signed and unsigned integers respectively which means that ptrdiff_t only has half the positive range of size_t and therefore doesn't have the capacity to index any allocatable array.

This is exactly my point. If there is going to be a switch replacing int with size_t, int is signed so its better to replace int with ptrdiff_t to avoid wrapping bugs. I am not worried about only having access to half of 2**64.

If I understand correctly, enabling the checks automatically means fixing all the current violations, and I do not believe we should prioritize fixing all the current violations. Though occasionally scrutinizing the violations to find current bugs might be useful. In general I believe "if it ain't broke don't fix it" since fixing it can often introduce bugs. 

You make a good point about memory capacity which is why I see a one-day use case for switching to ptrdiff_t.

**Reply by bcaddy** [reply]
Overflowing size_t is a bug and unlikely to occur due to its size. Underflowing size_t is also a bug and one that switching to ptrdiff_t won’t fix since negative indexes are a bug. So ptrdiff_t doesn’t solve any wrapping bugs. There are absolutely places when computing an index that you need a signed number and ptrdiff_t is that number but using it instead of size_t doesn’t fix any bugs

**Reply by alwinm** [reply]
> There are absolutely places when computing an index that you need a signed number and ptrdiff_t is that number but using it instead of size_t doesn’t fix any bugs

I am not sure I understand this sentence but I do believe you understand what I am conveying. Sometimes you need a signed number and ptrdiff_t is that number is all I'm trying to say and I do feel heard. The second part of the sentence confuses me because if you use ptrdiff_t instead of size_t where you need to, that prevents bugs tautologically because of the word "need." 

I think at this point we agree on the considerations that need to be taken into account for making this change and we can agree to disagree on the value/effort ratio.

**Reply by bcaddy** [reply]
I think the difference is that I’m not saying we should index with ptrdiff_t, just use it to compute the appropriate size_t variable

**Reply by alwinm** [reply]
Thank you for the clarification. If putting (size_t) in every index will fix both implicit widening and narrowing conversions without needing to change other variables currently declared as ```int``` I can see why that is appealing though I still am not convinced it's really worth prioritizing. 

I am mostly against changing variables currently explicitly declared as int to size_t except when the appropriate thought and consideration has been taken, and if necessary to use ptrdiff_t. But for this latter point I don't think you were proposing that, I felt I should just add a warning saying why we shouldn't do a blanket change from int to size_t and describing some possible traps along the way depending on the choices that are made.

**Reply by bcaddy** [reply]
Yeah, 100% not all `ints` should become `size_t`/`ptrdiff_t`, only where they're doing the job of indexing or related.

I think I'll put this on the back burner for now, for the same effort I think we can go through more checks and probably have a greater impact. If fancy new GPUs with like 500GB of memory come out we can make this a priority then.

**Reply by bcaddy** [reply]
Per a discussion with @evaneschneider offline I'm converting this into [an issue](https://github.com/cholla-hydro/cholla/issues/271) as it would probably be a good one for a new student or collaborator.
